# 机器学习工程师纳米学位
## 监督学习
## 项目 2: 搭建一个学生干预系统

欢迎来到机器学习工程师纳米学位的第二个项目！在此文件中，有些示例代码已经提供给你，但你还需要实现更多的功能让项目成功运行。除非有明确要求，你无须修改任何已给出的代码。以**'练习'**开始的标题表示接下来的代码部分中有你必须要实现的功能。每一部分都会有详细的指导，需要实现的部分也会在注释中以**'TODO'**标出。请仔细阅读所有的提示！

除了实现代码外，你还**必须**回答一些与项目和你的实现有关的问题。每一个需要你回答的问题都会以**'问题 X'**为标题。请仔细阅读每个问题，并且在问题后的**'回答'**文字框中写出完整的答案。我们将根据你对问题的回答和撰写代码所实现的功能来对你提交的项目进行评分。

>**提示：**Code 和 Markdown 区域可通过 **Shift + Enter** 快捷键运行。此外，Markdown可以通过双击进入编辑模式。

### 问题 1 - 分类 vs. 回归
*在这个项目中你的任务是找出那些如果不给予帮助，最重可能无法毕业的学生。你觉得这个问题是哪种类型的监督学习问题，是分类问题还是回归问题？为什么？*

**答案: **是分类问题，因为学校最终需要的结果是学生是否能毕业，结果只有‘yes’和‘no’这两个值，很明显是分类问题。

## 分析数据
运行下面区域的代码以载入学生数据集，以及一些此项目所需的Python库。注意数据集的最后一列`'passed'`是我们的预测的目标（表示学生是毕业了还是没有毕业），其他的列是每个学生的属性。

In [2]:
# 载入所需要的库
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# 载入学生数据集
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


### 练习: 分析数据
我们首先通过调查数据，以确定有多少学生的信息，并了解这些学生的毕业率。在下面的代码单元中，你需要完成如下的运算：
- 学生的总数， `n_students`。
- 每个学生的特征总数， `n_features`。
- 毕业的学生的数量， `n_passed`。
- 未毕业的学生的数量， `n_failed`。
- 班级的毕业率， `grad_rate`， 用百分数表示(%)。


In [7]:
print student_data.columns
# TODO： 计算学生的数量
n_students = len(student_data.index)

# TODO： 计算特征数量
n_features = len(student_data.columns)-1

# TODO： 计算通过的学生数
n_passed = dict(student_data['passed'].value_counts())["yes"]

# TODO： 计算未通过的学生数
n_failed = dict(student_data['passed'].value_counts())["no"]

# TODO： 计算通过率
grad_rate = float(n_passed)/float(n_students)*100

# 输出结果
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)

Index([u'school', u'sex', u'age', u'address', u'famsize', u'Pstatus', u'Medu',
       u'Fedu', u'Mjob', u'Fjob', u'reason', u'guardian', u'traveltime',
       u'studytime', u'failures', u'schoolsup', u'famsup', u'paid',
       u'activities', u'nursery', u'higher', u'internet', u'romantic',
       u'famrel', u'freetime', u'goout', u'Dalc', u'Walc', u'health',
       u'absences', u'passed'],
      dtype='object')
Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.09%


## 数据准备
在这个部分中，我们将要为建模、训练和测试准备数据
### 识别特征和目标列
你获取的数据中通常都会包含一些非数字的特征，这会导致一些问题，因为大多数的机器学习算法都会期望输入数字特征进行计算。

运行下面的代码单元将学生数据分成特征和目标列看一看他们中是否有非数字特征。

In [8]:
# 提取特征列
feature_cols = list(student_data.columns[:-1])

# 提取目标列 ‘passed’
target_col = student_data.columns[-1] 

# 显示列的列表
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# 将数据分割成特征数据和目标数据（即X_all 和 y_all）
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# 通过打印前5行显示特征信息
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### 预处理特征列

正如你所见，我们这里有几个非数值的列需要做一定的转换！它们中很多是简单的`yes`/`no`，比如`internet`。这些可以合理地转化为`1`/`0`（二元值，binary）值。

其他的列，如`Mjob`和`Fjob`，有两个以上的值，被称为_分类变量（categorical variables）_。处理这样的列的推荐方法是创建和可能值一样多的列（如：`Fjob_teacher`，`Fjob_other`，`Fjob_services`等），然后将其中一个的值设为`1`另外的设为`0`。

这些创建的列有时候叫做 _虚拟变量（dummy variables）_，我们将用[`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies)函数来完成这个转换。运行下面代码单元的代码来完成这里讨论的预处理步骤。

In [9]:
def preprocess_features(X):
    ''' 预处理学生数据，将非数字的二元特征转化成二元值（0或1），将分类的变量转换成虚拟变量
    '''
    
    # 初始化一个用于输出的DataFrame
    output = pd.DataFrame(index = X.index)

    # 查看数据的每一个特征列
    for col, col_data in X.iteritems():
        
        # 如果数据是非数字类型，将所有的yes/no替换成1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # 如果数据类型是类别的（categorical），将它转换成虚拟变量
        if col_data.dtype == object:
            # 例子: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # 收集转换后的列
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### 实现: 将数据分成训练集和测试集
现在我们已经将所有的 _分类的（categorical）_ 特征转换成数值了。下一步我们将把数据（包括特征和对应的标签数据）分割成训练集和测试集。在下面的代码单元中，你需要完成下列功能：
- 随机混洗（shuffle）切分数据(`X_all`, `y_all`) 为训练子集和测试子集。
  - 使用300个数据点作为训练集（约75%），使用95个数据点作为测试集（约25%）。
  - 如果可能的话，为你使用的函数设置一个`random_state`。
  - 将结果存储在`X_train`, `X_test`, `y_train`和 `y_test`中。

In [10]:
# TODO：在这里导入你可能需要使用的另外的功能
from sklearn.cross_validation import train_test_split
# TODO：设置训练集的数量
num_train = 300

# TODO：设置测试集的数量
num_test = X_all.shape[0] - num_train

# TODO：把数据集混洗和分割成上面定义的训练集和测试集
X_train ,X_test , y_train , y_test = train_test_split(X_all,y_all,test_size = num_test,random_state = 1)

# 显示分割的结果
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

Training set has 300 samples.
Testing set has 95 samples.


## 训练和评价模型
在这个部分，你将选择3个适合这个问题并且在`scikit-learn`中已有的监督学习的模型。首先你需要说明你选择这三个模型的原因，包括这些数据集有哪些特点，每个模型的优点和缺点各是什么。然后，你需要将这些模型用不同大小的训练集（100个数据点，200个数据点，300个数据点）进行训练，并用F<sub>1</sub>的值来衡量。你需要制作三个表，每个表要显示训练集大小，训练时间，预测时间，训练集上的F<sub>1</sub>值和测试集上的F<sub>1</sub>值（每个模型一个表）。

**这是目前** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **里有的监督学习模型，你可以从中选择:**
- Gaussian Naive Bayes (GaussianNB) 朴素贝叶斯
- Decision Trees 决策树
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM) 向量模型机
- Logistic Regression 逻辑回归

### 问题 2 - 应用模型
*列出三个适合这个问题的监督学习算法模型。每一个你选择的模型：*

- 描述一个该模型在真实世界的一个应用场景。（你需要为此做点研究，并给出你的引用出处）
- 这个模型的优势是什么？他什么情况下表现最好？
- 这个模型的缺点是什么？什么条件下它表现很差？
- 根据我们当前数据集的特点，为什么这个模型适合这个问题。

**回答: **
### 1.朴素贝叶斯模型
   - 在直接营销中的应用，通过利用朴素贝叶斯模型分析产品门店销售的聚类特征、客户购买事务数据集，得到客户对于产品的购买行为模式（来自论文《基于朴素贝叶斯分类器的个人信用评估模型》）。
   - 其优点为算法逻辑简单，易于实现，算法实施的时间空间开销小，算法性能稳定，对于不同特点的数据其分类性能差别不大。在各属性条件独立的情况下表现最好。
   - 其缺点为模型假设属性之间相互独立，这个假设在实际应用中往往是不成立的。若属性之间相关性大，模型表现很差。
   - 训练集维数高，样本数量相对比较小，适合使用高偏差/低方差的分类器，且数据集中的特征相关性不大，比较适合用朴素贝叶斯模型。
   
### 2.随机森林模型
   - 随机森林在人脸识别中的应用，通过随机森林模型分析脸部图像块来定位人脸的10个关键特征点（来自论文《Real-time facial feature detection using conditional regression forests》）
   - 随机森林几乎不需要输入准备，它可以处理二元特征，分类特征，数字特征，并且不需要任何缩放处理；可实现隐式特征选择；训练速度非常快；不容易过拟合。当样本数量少容易过拟合，特征难以筛选的情况下使用随机森林效果好。
   - 随机森林模型的缺点是它已经被证明在某些噪音较大的分类或回归问题上会过拟；对于有不同取值的特征的数据，取值划分较多的特征会对随机森林产生更大的影响，所以随机森林在这种数据上产出的特征权值是不可信的。在样本噪声过大或者有取值划分较多的特征的情况下它的表现很差。
   - 数据中label只有两个值，属于分类问题；数据集样本比较小，大部分特征为二元特征，且不确定是否线性可分，可使用随机森林模型。
   
### 3.SVM模型
   - SVM模型在图像分类中的应用，研究人员首先对图像进行分块，计算各个分块的人眼感知度权值，再使用了一种基于泛化的直方图相交核函数建立SVM模型。该模型能有效地对复杂的自然场景图像进行分类，且各项分类性能指标都优于其他算法。（来自论文《基于SVM的图像分类与标注方法的研究》）
   - SVM模型在解决小样本、非线性及高维模式识别中表现出许多特有的优势，可以避免神经网络结构选择和局部极小点问题。当数据非线性可分时用SVM模型效果很好。
   - SVM算法对大规模训练样本难以实施，当样本数目很大时该矩阵的存储和计算将耗费大量的机器内存和运算时间；经典的支持向量机算法只给出了二类分类的算法，不适合单独使用SVM模型解决多分类问题；对缺失数据敏感。当样本数量多或者并非二分类的时候使用SVM效果差。
   - 数据中label只有两个值，属于分类问题，数据集样本比较小，且不确定是否线性可分，可使用SVM模型。


### 准备
运行下面的代码单元以初始化三个帮助函数，这三个函数将能够帮你训练和测试你上面所选择的三个监督学习算法。这些函数是：
- `train_classifier` - 输入一个分类器和训练集，用数据来训练这个分类器。
- `predict_labels` - 输入一个训练好的分类器、特征以及一个目标标签，这个函数将帮你做预测并给出F<sub>1</sub>的值.
- `train_predict` - 输入一个分类器以及训练集和测试集，它可以运行`train_clasifier`和`predict_labels`.
 - 这个函数将分别输出训练集的F<sub>1</sub>值和测试集的F<sub>1</sub>值

In [11]:
def train_classifier(clf, X_train, y_train):
    ''' 用训练集训练分类器 '''
    
    # 开始计时，训练分类器，然后停止计时
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' 用训练好的分类器做预测并输出F1值'''
    
    # 开始计时，作出预测，然后停止计时
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # 输出并返回结果
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' 用一个分类器训练和预测，并输出F1值 '''
    
    # 输出分类器名称和训练集大小
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # 训练一个分类器
    train_classifier(clf, X_train, y_train)
    
    # 输出训练和测试的预测结果
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))

### 练习: 模型评价指标
借助于上面定义的函数，你现在需要导入三个你选择的监督学习模型，然后为每一个模型运行`train_predict`函数。请记住，对于每一个模型你需要在不同大小的训练集（100，200和300）上进行训练和测试。所以，你在下面应该会有9个不同的输出（每个模型都有训练集大小不同的三个输出）。在接下来的代码单元中，你将需要实现以下功能：
- 引入三个你在上面讨论过的监督式学习算法模型。
- 初始化三个模型并将它们存储在`clf_A`， `clf_B` 和 `clf_C`中。
 - 如果可能对每一个模型都设置一个`random_state`。
 - **注意:** 这里先使用每一个模型的默认参数，在接下来的部分中你将需要对某一个模型的参数进行调整。
- 创建不同大小的训练集用来训练每一个模型。
 - *不要再混洗和再分割数据！新的训练集要取自`X_train`和`y_train`.*
- 对于每一个模型要用不同大小的训练集来训练它，然后在测试集上做测试（总共需要9次训练测试）   
**注意:** 在下面的代码单元后面我们提供了三个表用来存储你的结果。

In [12]:
# TODO：从sklearn中引入三个监督学习模型
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn import tree

# TODO：初始化三个模型
clf_A = GaussianNB()
clf_B = RandomForestClassifier(random_state = 1)
clf_C = SVC(random_state = 1)
#clf_D = LogisticRegression()
#clf_E = tree.DecisionTreeClassifier()

# TODO：设置训练集大小
X_train_100 = X_train[:100]
y_train_100 = y_train[:100]

X_train_200 = X_train[:200]
y_train_200 = y_train[:200]

X_train_300 = X_train[:300]
y_train_300 = y_train[:300]

# TODO：对每一个分类器和每一个训练集大小运行'train_predict' 
train_predict(clf_A, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_B, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_C, X_train_100, y_train_100, X_test, y_test)
#train_predict(clf_D, X_train_100, y_train_100, X_test, y_test)
#train_predict(clf_E, X_train_100, y_train_100, X_test, y_test)
print "======================================================="
train_predict(clf_A, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_B, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_C, X_train_200, y_train_200, X_test, y_test)
#train_predict(clf_D, X_train_200, y_train_200, X_test, y_test)
#train_predict(clf_E, X_train_200, y_train_200, X_test, y_test)
print "======================================================="
train_predict(clf_A, X_train_300, y_train_300, X_test, y_test)
train_predict(clf_B, X_train_300, y_train_300, X_test, y_test)
train_predict(clf_C, X_train_300, y_train_300, X_test, y_test)
#train_predict(clf_D, X_train_300, y_train_300, X_test, y_test)
#train_predict(clf_E, X_train_300, y_train_300, X_test, y_test)

Training a GaussianNB using a training set size of 100. . .
Trained model in 0.0190 seconds
Made predictions in 0.0000 seconds.
F1 score for training set: 0.8346.
Made predictions in 0.0000 seconds.
F1 score for test set: 0.7402.
Training a RandomForestClassifier using a training set size of 100. . .
Trained model in 0.0400 seconds
Made predictions in 0.0010 seconds.
F1 score for training set: 0.9767.
Made predictions in 0.0000 seconds.
F1 score for test set: 0.7313.
Training a SVC using a training set size of 100. . .
Trained model in 0.0920 seconds
Made predictions in 0.0010 seconds.
F1 score for training set: 0.8591.
Made predictions in 0.0000 seconds.
F1 score for test set: 0.8333.
Training a GaussianNB using a training set size of 200. . .
Trained model in 0.0010 seconds
Made predictions in 0.0000 seconds.
F1 score for training set: 0.7879.
Made predictions in 0.0010 seconds.
F1 score for test set: 0.6446.
Training a RandomForestClassifier using a training set size of 200. . .
Tra

### 结果表格
编辑下面的表格看看在[Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables)中如何设计一个表格。你需要把上面的结果记录在表格中。

** 分类器 1 朴素贝叶斯模型 - ?**  

| 训练集大小 | 训练时间 | 预测时间 (测试) | F1值 (训练) | F1值 (测试) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               | 0.0010              | 0.0010              |      0.8346            | 0.7402           |
| 200               | 0.0010              | 0.0010              |      0.7879            | 0.6446           | 
| 300               | 0.0010              | 0.0000              |      0.7921            | 0.6720           |

** 分类器 2 随机森林模型 - ?**  

| 训练集大小 | 训练时间 | 预测时间 (测试) | F1值 (训练) | F1值 (测试) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               | 0.0300              | 0.0010              |      0.9767            | 0.7313           |
| 200               | 0.0340              | 0.0010              |      0.9925            | 0.7519           | 
| 300               | 0.0330              | 0.0020              |      0.9924            | 0.7970           |

** 分类器 3 SVM模型 - ?**  

| 训练集大小 | 训练时间 | 预测时间 (测试) | F1值 (训练) | F1值 (测试) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               | 0.0010              | 0.0010              |      0.8591            | 0.8333           |
| 200               | 0.0040              | 0.0030              |      0.8581            | 0.8408           | 
| 300               | 0.0080              | 0.0020              |      0.8584            | 0.8462           |

## 选择最佳模型
在最后这一部分中，你将从三个监督学习模型中选择一个用在学生数据上的最佳模型。然后你将在最佳模型上用全部的训练集（`X_train`和`y_train`）运行一个网格搜索算法，在这个过程中，你要至少调整一个参数以提高模型的F<sub>1</sub>值（相比于没有调参的模型的分值有所提高）。 

### 问题 3 - 选择最佳模型
*给予你上面做的实验，用一到两段话，向（学校）监事会解释你将选择哪个模型作为最佳的模型。哪个模型在现有的数据，有限的资源、开支和模型表现综合来看是最好的选择？*

**回答: **我会使用SVM模型，学校希望通过分析数据来预测学生是否可以毕业，这属于二分类问题，使用分类模型可以很好的解决。因此我从中选择了3个优秀的分类模型，朴素贝叶斯模型、随机森林模型、SVM模型对学生数据进行分析。发现模型的训练分数和测试分数均在0.83到0.86之间，说明模型很好的防止了过拟合。而且在3个模型中，SVM模型得出最高的f1分数（0.8462分），这个分数说明该模型能很好的预测学生是否毕业。虽然SVM模型比起其他两个模型需要耗费大量的机器内存和运算时间。但是因为所需要分析的样本数量并不多（不到1000），分析样本所花的时间皆在毫秒级别。所以使用SVM模型消耗的资源在可接受范围之类。综上所述，SVM模型是处理该问题的最好的模型。

### 问题 4 - 用通俗的语言解释模型
*用一到两段话，向（学校）监事会用外行也听得懂的话来解释最终模型是如何工作的。你需要解释所选模型的主要特点。例如，这个模型是怎样被训练的，它又是如何做出预测的。避免使用高级的数学或技术术语，不要使用公式或特定的算法名词。*

**回答: **
我选中的模型最终采用如下方法进行工作的，它主要分为两个步骤。
1、训练数据
   训练数据简单的来说就是找到一个方法，将毕业的学生和不能毕业的学生区分开来。每个学生样本都包含着大量的信息，而SVM模型使用特殊的方法将每个学生样本都形象使用空间上的一个点来表示，而这个点在空中的位置信息与该学生的所有信息相对应着。所以，如何训练数据的问题就变为如何将空间上395个点划分成两类点，一类点代表着所有毕业的学生，一类点代表着所有不能毕业的学生。通常，同一类点所包含的信息相近，因此在空间中的位置也比较相近，所以同一类点大都会聚集成一团。而这时，SVM模型就通过分析这些点的位置，找出一个平面，使一类点在平面的一侧，一类点在平面的另一侧（如果无法找到这样的平面，模型就会改进之前样本转换成点地址信息的方法，重新构建出一个空间，直到位于该空间中的点可以被一个平面划分开）。当然，并不是所有点都要分析到，比如说远离跟自己不同属一类点的点，这些点对找出该平面几乎没有什么帮助，就不需要分析；相反，对于十分接近跟自己不同属一类点的点，它们对找出该平面具有决定性的作用。而这些点的所代表的地址信息也被称为支持向量。
2、预测数据
   当我们找出这个平面之后，我们就可以预测数据了。首先我们要将要预测的学生样本信息按照之前特殊的方法转变成一个点在空间上的地址信息，使其形象的在空间中表现出来。然后观察这个点在之前SMV模型找出的平面的哪一侧，如果它跟毕业学生那一类点是同一侧，SVM模型就预测这个学生会毕业，反之亦然。这就是SVM模型预测数据的过程。

### 练习: 模型调参
细调选择的模型的参数。使用网格搜索（`GridSearchCV`）来至少调整模型的重要参数（至少调整一个），这个参数至少需给出并尝试3个不同的值。你要使用整个训练集来完成这个过程。在接下来的代码单元中，你需要实现以下功能：
- 导入 [`sklearn.grid_search.gridSearchCV`](http://scikit-learn.org/stable/modules/generated/sklearn.grid_search.GridSearchCV.html) 和 [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- 创建一个对于这个模型你希望调整参数的字典。
 - 例如: `parameters = {'parameter' : [list of values]}`。
- 初始化你选择的分类器，并将其存储在`clf`中。
- 使用`make_scorer` 创建F<sub>1</sub>评分函数并将其存储在`f1_scorer`中。
 - 需正确设定参数`pos_label`的值！
- 在分类器`clf`上用`f1_scorer` 作为评价函数运行网格搜索,并将结果存储在`grid_obj`中。
- 用训练集(`X_train`, `y_train`)训练grid search object,并将结果存储在`grid_obj`中。

In [19]:
# TODO: 导入 'GridSearchCV' 和 'make_scorer'
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import make_scorer,f1_score
# TODO：创建你希望调整的参数列表

parameters = {'C': [0.001,0.01,0.1,1,10],'kernel':["linear","rbf","poly","sigmoid"],'gamma': [0.001,0.01,0.1,1,10,100,1000],'degree':[3,4,5]}

# TODO：初始化分类器
clf = SVC(random_state = 1)

# TODO：用'make_scorer'创建一个f1评分函数
f1_scorer = make_scorer(f1_score,pos_label = 'yes')

# TODO：在分类器上使用f1_scorer作为评分函数运行网格搜索
grid_obj = GridSearchCV(clf,parameters,scoring = f1_scorer)

# TODO: Fit the grid search object to the training data and find the optimal parameters
# TODO：用训练集训练grid search object来寻找最佳参数
grid_obj.fit(X_train,y_train)

# Get the estimator
# 得到预测的结果
clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
# 输出经过调参之后的训练集和测试集的F1值
print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))

Made predictions in 0.0040 seconds.
Tuned model has a training F1 score of 0.9754.
Made predictions in 0.0000 seconds.
Tuned model has a testing F1 score of 0.8481.


### 问题 5 - 最终的 F<sub>1</sub> 值
*最终模型的训练和测试的F<sub>1</sub>值是多少？这个值相比于没有调整过参数的模型怎么样？*

**回答: **最终模型训练的F1值是0.9754，测试的F1值是0.8481。这些值比使用默认参数的模型要高。

> **注意**: 当你写完了所有的代码，并且回答了所有的问题。你就可以把你的 iPython Notebook 导出成 HTML 文件。你可以在菜单栏，这样导出**File -> Download as -> HTML (.html)**把这个 HTML 和这个 iPython notebook 一起做为你的作业提交。  